2.train_model/scripts/nbconverted/train_single_class_models.py

```diff
@@ -0,0 +1,154 @@
+    GridSearchCV,
+)
+from sklearn.utils import shuffle, parallel_backend
+from sklearn.exceptions import ConvergenceWarning
+from joblib import dump
+
+import sys
+
+sys.path.append("../utils")
+from split_utils import get_features_data
+from train_utils import get_dataset, get_X_y_data
+
+
+# In[2]:
+
+
+# set numpy seed to make random operations reproduceable
+np.random.seed(0)
+
+# load training data from indexes and features dataframe
+data_split_path = pathlib.Path(f"../1.split_data/indexes/data_split_indexes.tsv")
+features_dataframe_path = pathlib.Path("../0.download_data/data/labeled_data.csv.gz")
+
+# dataframe with only the labeled data we want (exclude certain phenotypic classes)
+features_dataframe = get_features_data(features_dataframe_path)
+data_split_indexes = pd.read_csv(data_split_path, sep="\t", index_col=0)
+
+# get training data from labeled data
+training_data = get_dataset(features_dataframe, data_split_indexes, "train")
+# training_data
+
+
+# In[3]:
+
+
+# specify model types and feature types
+model_types = ["final", "shuffled_baseline"]
+feature_types = ["CP", "DP", "CP_and_DP"]
+
+# create stratified data sets for k-fold cross validation
+straified_k_folds = StratifiedKFold(n_splits=10, shuffle=False)
+
+# create logistic regression model with following parameters
+log_reg_model = LogisticRegression(
+    penalty="elasticnet",
+    solver="saga",
+    max_iter=100,
+    n_jobs=-1,
+    random_state=0,
+    class_weight="balanced"
+)
+
+# specify parameters to tune for
+parameters = {"C": np.logspace(-3, 3, 7), "l1_ratio": np.linspace(0, 1, 11)}
+print(f"Parameters being tested during grid search: {parameters}\n")
+
+# create grid search with cross validation with hypertuning params
+grid_search_cv = GridSearchCV(
+    log_reg_model, parameters, cv=straified_k_folds, n_jobs=-1, scoring="f1_weighted"
+)
+
+# train model on each combination of model type and feature type
+for model_type in model_types:
+    for feature_type in feature_types:
+        for phenotypic_class in training_data["Mitocheck_Phenotypic_Class"].unique():
+            # create results directory
+            results_dir = pathlib.Path(
+                f"models/single_class_models/{phenotypic_class}_models/"
+            )
+            results_dir.mkdir(parents=True, exist_ok=True)
+
+            # get number of labels for this specific phenotypic_class
+            phenotypic_class_counts = (
+                training_data.loc[
+                    training_data["Mitocheck_Phenotypic_Class"] == phenotypic_class
+                ]
+            ).shape[0]
+            print(
+                f"Training {model_type} model on {feature_type} features for {phenotypic_class} phenotypic class with {phenotypic_class_counts} positive labels..."
+            )
+
+            # create deep copy of training data so we can make modifications without affecting original training data
+            class_training_data = training_data.copy(deep=True)
```

Comment:
Data duplication can be problematic if the dataset is large. Do you think adding a column to the dataframe would solve this issue or storing the edited labels into a list and then concatenating it to the original dataframe.